File: binary_search_tree.py
class Node:
    
    def __init__(self,data):
        self.data=data
        self.left=None
        self.right=None
    
    def insert(self, data):
        if self.data>data:
            if self.left==None:
                self.left=Node(data)
            else:
                self.left.insert(data)
        if self.data<data:
            if self.right==None:
                self.right=Node(data)
            else:
                self.right.insert(data)

class BinarySearchTree:
    
    def __init__(self):
        self.root=None
    
    def insert(self,data):
        if self.root==None:
            self.root=Node(data)
        else:
            self.root.insert(data)
    
    def search(self,item):
        tmp=self.root
        while tmp!=None:
            if tmp.data<item:
                tmp=tmp.right
            elif tmp.data>item:
                tmp=tmp.left
            else:
                return True
        return False

File: src/app/binarysearchtree/BinarySearchTree.java
package app.binarysearchtree;

public class BinarySearchTree<E extends Comparable<E>>
{
    private Node<E> root;
    public void insert(E data)
    {
        if(root==null)
            root=new Node<>(data);
        else
            root.insert(data);
    }
    public boolean search(E data)
    {
        if(root==null)
            return false;
        Node<E> tmp=root;
        while(true)
        {
            int result=data.compareTo(tmp.getData());
            if(result>0)
            {
                if(tmp.hasRight())
                    tmp=tmp.getRight();
                else
                    return false;
            }
            else if(result<0)
            {
                if(tmp.hasLeft())
                    tmp=tmp.getLeft();
                else
                    return false;
            }
            else
                return true;
        }
    }
    private class Node<T extends Comparable<T>>
    {
        private final T data;
        private Node<T> left,right;
        public Node(T data)
        {
            this.data=data;
        }
        public void insert(T data)
        {
            if(data.compareTo(this.data)>0)
            {
                if(right==null)
                    right=new Node<>(data);
                else
                    right.insert(data);
            }
            else
            {
                if(left==null)
                    left=new Node<>(data);
                else
                    left.insert(data);
            }
        }
        public T getData()
        {
            return data;
        }
        public Node<T> getLeft()
        {
            return left;
        }
        public Node<T> getRight()
        {
            return right;
        }
        public boolean hasRight()
        {
            return right!=null;
        }
        public boolean hasLeft()
        {
            return left!=null;
        }
    }
}